keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,582 @@
+
+#### Memory Maps at start-up (with examples)
+
+The figure below illustrates `Memory Maps` (for regular memory) shortly after node start-up.
+
+![alt text](./mem_tables_startup_without_naf.png)
+
+The figure below illustrates `Memory Maps` (for regular memory) shortly after node start-up and when [Node Allocatable Feature](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/) is used, i.e. the value of `SystemReserved` must be configured by the node administrator as for example:
+
+`--reserved-memory={numa-node=0, type=memory, limit=500MB}, {numa-node=1,type=memory, limit=100MB}`
+
+![alt text](./mem_tables_startup.png)
+
+The value of `SystemReserved` is pre-configured via our new `kubelet` flag; therefore, `SystemReserved` remains constant during the runtime. The value of `SystemReserved` directly implies the value of `Allocatable`, so `Allocatable` is also invariable during the runtime. 
+
+`SystemReserved` indicates the amount of memory which is pre-reserved, i.e. **guaranteed**, for the system, i.e., kernel, OS daemons, and core node components such as *kubelet* (*kubelet* deamons).
+
+#### Memory Maps at runtime (with examples)
+
+The figure below illustrates `Memory Maps` (for regular memory) during the node runtime. The figure shows the situation after a successful deployment of container _"A"_ and container _"B"_. Also, *Memory Manager* **guaranteed** (reserved) memory for the containers. In the example, the containers consume less memory than it was reserved (**guaranteed**), so for both containers their memory consumption remains below their memory limits (e.g. `cgroup` memory limit), and containers operate as normal.
+
+![alt text](./mem_tables_runtime.png)
+
+However, if **Out-of-Memory** (OOM) situation occurs, i.e., a particular memory limit is reached and crossed (e.g. by the containers) within _K8S_ node, the node triggers one of [the mechanisms](#kubernetes-nodes-memory-management-mechanisms-and-their-relation-to-the-memory-manager). Depending on what a particular limit is crossed (`cgroups` memory limit, `hard-eviction-treshold` memory limit, etc.), a suitable mechanism becomes triggered accordingly. As already noticed, such an OOM situation is handled by *kubelet* or system (OOM kernel killer) and not by *Memory Manager* itself. 
+
+### Simulation - how the Memory Manager works? (by examples)
+
+In this section, a number of slides is intended to illustrate the operation of the Memory Manager in various situations, i.e., to show how the pods in different QoS classes (Guaranteed, BestEffort/Burstable) are handled by the Memory Manager or other means, and how the disjoint groups of NUMA nodes are dynamically managed (created or removed) by the Memory Manager.  
+
+![alt text](./mm_simulation.jpg)
```

Comment:
It's hard to comment on these individual slides since they are not inline with the text. In any case.

*For the slide titled "Reject Pod2":*
The way it's presented here makes it seems strange that you wouldn't want to allocate memory from this group (since the actual memory allocated to the pod would in fact come from a single NUMA node in this example). The issue would be better demonstrated if the 5GB to service this pod was only available as (e.g.) 2GB from node 0 and 3GB from node 1.  This should clearly be rejected because the memory is coming from 2 different NUMA nodes, when waiting to service the pod frtom a different node (or waiting until this node emptied out) would let it be serviced from just 1 NUMA node. Not sure how or if you want to clear this up, but that is the main reasoning behind this.

Reply:
The slide _"Reject Pod2"_ looks how it looks because it depicts a standard procedure, i.e.:
1. Memory Manager finds out from which NUMA nodes a pod should allocate memory, i.e. Pod2 must use both NUMA nodes 
2. Memory Manager updates `cpuset.mems`. 
3. Kernel uses its default NUMA Memory Policy, i.e. "local" [1].
4. Memory is allocated "locally", i.e. from the NUMA node where CPUs are pinned. 
5. As normal, CPU Manager pinned CPUs to the first NUMA node.
6. Memory is consumed from the first NUMA node completely (10G). 
7. Then, container consumes memory from the second NUMA node (5G).

Please, correct me if it operates in a different way.

[1] https://www.systutorials.com/docs/linux/man/2-set_mempolicy/

A comment was added below the slides:

> \#\#\#\# Slide "Reject Pod2"
> 
> The slide _"Reject Pod2"_ results from a standard scenario, i.e.:
> 
> 0. Both NUMA nodes are empty.
> 1. Memory Manager finds out from which NUMA nodes a pod should allocate memory, i.e. *Pod2* must use both NUMA nodes. 
> 2. Memory Manager updates `cpuset.mems`. 
> 3. Kernel uses its default #(NUMA Memory Policy)(https://www.systutorials.com/docs/linux/man/2-set_mempolicy/), i.e. "local".
> 4. Memory is allocated "locally", i.e. from the NUMA node where CPUs are pinned. 
> 5. As normal, CPU Manager pinned CPUs to the first NUMA node.
> 6. Therefore, memory is consumed from the first NUMA node completely (10G). 
> 7. And then, *Pod2* consumes memory from the second NUMA node (5G).
> 
> This might be surprising that *Pod2* is rejected although there is still enough spare memory at the second NUMA node. Nevertheless, the behaviour is correct and complies with both designs, i.e., disjoint NUMA node groups design and Topology Manager design.  
> 
> The issue in the slide could also be illustrated in a different way, i.e., 5GB available to *Pod2* could be available from both nodes (e.g. 2GB from the first and 3GB from the second NUMA node). This would show that the memory needs to be allocated from a couple of different NUMA nodes, and that *Pod2* admission should be clearly rejected.  
>